ddtrace/tracer/sqlcomment.go

```diff
@@ -0,0 +1,174 @@
+const (
+	SamplingPrioritySQLCommentKey   = "ddsp"
+	TraceIDSQLCommentKey            = "ddtid"
+	SpanIDSQLCommentKey             = "ddsid"
+	ServiceNameSQLCommentKey        = "ddsn"
+	ServiceVersionSQLCommentKey     = "ddsv"
+	ServiceEnvironmentSQLCommentKey = "dde"
+)
+
+// SQLCommentCarrier is a carrier implementation that injects a span context in a SQL query in the form
+// of a sqlcommenter formatted comment prepended to the original query text.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+type SQLCommentCarrier struct {
+	Query  string
+	Mode   SQLCommentInjectionMode
+	SpanID uint64
+}
+
+// NewSQLCommentCarrier returns a new instance of a SQLCommentCarrier
+func NewSQLCommentCarrier(query string, mode SQLCommentInjectionMode) (c *SQLCommentCarrier) {
+	c = new(SQLCommentCarrier)
+	c.Mode = mode
+	c.Query = query
+	c.SpanID = random.Uint64()
+	return c
+}
+
+// Inject injects a span context in the carrier's query.
+func (c *SQLCommentCarrier) Inject(spanCtx ddtrace.SpanContext) error {
+	if c.Mode == CommentInjectionDisabled {
+		return nil
+	}
+
+	tags := make(map[string]string)
+	if c.Mode == ServiceTagsInjection || c.Mode == FullSQLCommentInjection {
+		ctx, ok := spanCtx.(*spanContext)
+		var env, version string
+		if ok {
+			if e, ok := ctx.meta(ext.Environment); ok {
+				env = e
+			}
+			if v, ok := ctx.meta(ext.Version); ok {
+				version = v
+			}
+		}
+		if globalconfig.ServiceName() != "" {
+			tags[ServiceNameSQLCommentKey] = globalconfig.ServiceName()
+		}
+		if env != "" {
+			tags[ServiceEnvironmentSQLCommentKey] = env
+		}
+		if version != "" {
+			tags[ServiceVersionSQLCommentKey] = version
+		}
+	}
+	if c.Mode == FullSQLCommentInjection {
+		samplingPriority := 0
+
+		var traceID uint64
+		ctx, ok := spanCtx.(*spanContext)
+		if ok {
+			if sp, ok := ctx.samplingPriority(); ok {
+				samplingPriority = sp
+			}
+			if ctx.TraceID() > 0 {
+				traceID = ctx.TraceID()
+			}
+		}
+		if traceID == 0 {
+			traceID = c.SpanID
+		}
+		tags[TraceIDSQLCommentKey] = strconv.FormatUint(traceID, 10)
+		tags[SpanIDSQLCommentKey] = strconv.FormatUint(c.SpanID, 10)
+		tags[SamplingPrioritySQLCommentKey] = strconv.Itoa(samplingPriority)
+	}
+
+	c.Query = commentQuery(c.Query, tags)
+	return nil
+}
+
+// commentQuery returns the given query with the tags from the SQLCommentCarrier applied to it as a
+// prepended SQL comment. The format of the comment follows the sqlcommenter spec.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+func commentQuery(query string, tags map[string]string) string {
+	c := serializeTags(tags)
+	if c == "" {
+		return query
+	}
+	if query == "" {
+		return c
+	}
+	return fmt.Sprintf("%s %s", c, query)
+}
+
+func serializeTags(tags map[string]string) (comment string) {
+	if len(tags) == 0 {
+		return ""
+	}
+	serializedTags := make([]string, 0, len(tags))
+	for k, v := range tags {
+		serializedTags = append(serializedTags, serializeTag(k, v))
```

Comment:
Can't we just inline all the code into this one single function? This seems a bit complex to me. Why do we need all the other smaller bits of functions which are not re-used anywhere and barely have a bit of code in them. It scatters the attention and makes it hard to follow. In this case I think verbosity is a win. You get to read the whole thing in one function and it's all clear using Go standard library call, without any newly introduced concepts (functions) that one needs to navigate to a different line to study in order to piece it all together. I hope it makes sense.